plugins/Marketplace/Api/Client.php

```diff
@@ -0,0 +1,253 @@
+<?php
+/**
+ * Piwik - free/libre analytics platform
+ *
+ * @link http://piwik.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+namespace Piwik\Plugins\Marketplace\Api;
+
+use Piwik\Cache;
+use Piwik\Common;
+use Piwik\Container\StaticContainer;
+use Piwik\Filesystem;
+use Piwik\Http;
+use Piwik\Plugin;
+use Piwik\Plugins\Marketplace\Api\Service;
+use Piwik\SettingsServer;
+use Piwik\Version;
+use Exception as PhpException;
+use Psr\Log\LoggerInterface;
+
+/**
+ *
+ */
+class Client
+{
+    const CACHE_TIMEOUT_IN_SECONDS = 1200;
+    const HTTP_REQUEST_TIMEOUT = 60;
+
+    /**
+     * @var Service
+     */
+    private $service;
+
+    /**
+     * @var Cache\Lazy
+     */
+    private $cache;
+
+    /**
+     * @var Plugin\Manager
+     */
+    private $pluginManager;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    public function __construct(Service $service, Cache\Lazy $cache, LoggerInterface $logger)
+    {
+        $this->service = $service;
+        $this->cache = $cache;
+        $this->logger = $logger;
+        $this->pluginManager = Plugin\Manager::getInstance();
+    }
+
+    public function getPluginInfo($name)
+    {
+        $action = sprintf('plugins/%s/info', $name);
+
+        return $this->fetch($action, array());
+    }
+
+    public function getConsumer()
+    {
+        try {
+            $consumer = $this->fetch('consumer', array());
+        } catch (Exception $e) {
+            $consumer = null;
+        }
+
+        return $consumer;
+    }
+
+    private function getRandomTmpPluginDownloadFilename()
+    {
+        $tmpPluginPath = StaticContainer::get('path.tmp') . '/latest/plugins/';
+
+        // we generate a random unique id as filename to prevent any user could possibly download zip directly by
+        // opening $piwikDomain/tmp/latest/plugins/$pluginName.zip in the browser. Instead we make it harder here
+        // and try to make sure to delete file in case of any error.
+        $tmpPluginFolder = Common::generateUniqId();
+
+        return $tmpPluginPath . $tmpPluginFolder . '.zip';
+    }
+
+    public function download($pluginOrThemeName)
+    {
+        @ignore_user_abort(true);
```

Comment:
Using such methods like ignore user abort and max execution time in such a part is rather not good but decided to put it here for now to make sure we won't forget to call it. But can also move it out of this method and make sure it's there wherever this method is called